pkg/daemon/status.go

```diff
@@ -0,0 +1,142 @@
+package daemon
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	sriovnetworkv1 "github.com/k8snetworkplumbingwg/sriov-network-operator/api/v1"
+	"github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/consts"
+	"github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/vars"
+)
+
+const (
+	Unknown = "Unknown"
+)
+
+func (dn *DaemonReconcile) updateSyncState(ctx context.Context, desiredNodeState *sriovnetworkv1.SriovNetworkNodeState, status, failedMessage string) error {
+	funcLog := log.Log.WithName("updateSyncState")
+	currentNodeState := &sriovnetworkv1.SriovNetworkNodeState{}
+	desiredNodeState.Status.SyncStatus = status
+	desiredNodeState.Status.LastSyncError = failedMessage
+
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		if err := dn.client.Get(ctx, client.ObjectKey{desiredNodeState.Namespace, desiredNodeState.Name}, currentNodeState); err != nil {
+			funcLog.Error(err, "failed to get latest node state",
+				"SyncStatus", status,
+				"LastSyncError", failedMessage)
+			return err
+		}
+
+		funcLog.V(2).Info("update nodeState status",
+			"CurrentSyncStatus", currentNodeState.Status.SyncStatus,
+			"CurrentLastSyncError", currentNodeState.Status.LastSyncError,
+			"NewSyncStatus", status,
+			"NewFailedMessage", failedMessage)
+
+		err := dn.client.Status().Patch(ctx, desiredNodeState, client.MergeFrom(currentNodeState))
+		if err != nil {
+			funcLog.Error(err, "failed to update node state status",
+				"SyncStatus", status,
+				"LastSyncError", failedMessage)
+			return err
+		}
+		return nil
+	})
+
+	if retryErr != nil {
+		funcLog.Error(retryErr, "failed to update node state status")
+		return retryErr
+	}
+
+	dn.recordStatusChangeEvent(ctx, currentNodeState.Status.SyncStatus, status, failedMessage)
+	return nil
+}
+
+func (dn *DaemonReconcile) shouldUpdateStatus(current, desiredNodeState *sriovnetworkv1.SriovNetworkNodeState) (bool, error) {
+	// check number of interfaces are equal
+	if len(current.Status.Interfaces) != len(desiredNodeState.Status.Interfaces) {
+		return true, nil
+	}
+
+	// check for bridges
+	if !reflect.DeepEqual(current.Status.Bridges, desiredNodeState.Status.Bridges) {
+		return true, nil
+	}
+
+	// check for system
+	if !reflect.DeepEqual(current.Status.System, desiredNodeState.Status.System) {
+		return true, nil
+	}
+
+	// check for interfaces
+	// we can't use deep equal here because if we have a vf inside a pod is name will not be available for example
+	// we use the index for both lists
+	c := current.Status.DeepCopy().Interfaces
+	d := desiredNodeState.Status.DeepCopy().Interfaces
+	for idx := range d {
+		// check if it's a new device
```

Comment:
```suggestion
        if idx >= len(c) {
          return true, nil
        }
		// check if it's a new device
```

I don't know if it can ever happen (maybe if a new device is hot-plugged between reconcile loops). But it's safer to check for slice length

Reply:
but in theory (not sure if it's possible) you can have a new device like hot-unplug and after that hot-plug so the len will be the same but the device is not no?

Reply:
in both case, the `if idx >= len(c) {` can save a nil dereferencing. right?